File: src/main/java/no/fintlabs/kafka/topic/name/TopicNameParameterCharacterValidationService.java
package no.fintlabs.kafka.topic.name;

import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Objects;

@Service
class TopicNameParameterCharacterValidationService {

    void validateValueCharacters(String parameterName, Collection<String> values) {
        values.forEach(value -> validateValueCharacters(parameterName, value));
    }

    void validateValueCharacters(String parameterName, String value) {
        if (Objects.isNull(value)) {
            throw new IllegalArgumentException("Pattern value cannot be null");
        }
        String ILLEGAL_TOPIC_CHARACTER_INFO = "Topic components cannot include '.' or uppercase letters.";
        if (value.contains(".")) {
            throw new IllegalArgumentException(String.format(
                    "Parameter'%s' contains '.'. %s",
                    optionalParameterNameInsert(parameterName),
                    ILLEGAL_TOPIC_CHARACTER_INFO
            ));
        }
        if (value.chars().anyMatch(Character::isUpperCase)) {
            throw new IllegalArgumentException(String.format(
                    "Parameter'%s' contains uppercase letter(s). %s",
                    optionalParameterNameInsert(parameterName),
                    ILLEGAL_TOPIC_CHARACTER_INFO
            ));
        }
    }

    private String optionalParameterNameInsert(String parameterName) {
        return Objects.isNull(parameterName) ? "" : " " + parameterName;
    }

}

File: src/main/kotlin/no/novari/kafka/topic/name/TopicNameParameterCharacterValidationService.kt
package no.novari.kafka.topic.name

import org.springframework.stereotype.Service

private const val ILLEGAL_TOPIC_CHARACTER_INFO = "Topic components cannot include '.' or uppercase letters."

@Service
internal class TopicNameParameterCharacterValidationService {
    fun validateValueCharacters(
        parameterName: String?,
        values: Collection<String>,
    ) {
        values.forEach { validateValueCharacters(parameterName, it) }
    }

    fun validateValueCharacters(
        parameterName: String?,
        value: String?,
    ) {
        requireNotNull(value) { "Pattern value cannot be null" }
        require(!value.contains(".")) {
            "Parameter${optionalParameterNameInsert(parameterName)} contains '.'. $ILLEGAL_TOPIC_CHARACTER_INFO"
        }
        require(value.none { it.isUpperCase() }) {
            "Parameter${optionalParameterNameInsert(parameterName)} contains uppercase letter(s). " +
                ILLEGAL_TOPIC_CHARACTER_INFO
        }
    }

    private fun optionalParameterNameInsert(parameterName: String?): String =
        if (parameterName == null) "" else " '$parameterName'"
}
